networkit/cpp/community/ParallelLeiden.cpp

```diff
@@ -0,0 +1,534 @@
+#include <networkit/community/ParallelLeiden.hpp>
+
+namespace NetworKit {
+ParallelLeiden::ParallelLeiden(const Graph &graph, int iterations, bool randomize, double gamma)
+    : CommunityDetectionAlgorithm(graph), gamma(gamma), numberOfIterations(iterations),
+      random(randomize) {
+    this->result = Partition(graph.numberOfNodes());
+    this->result.allToSingletons();
+}
+
+void ParallelLeiden::run() {
+    auto totalTime = Aux::Timer();
+    totalTime.start();
+    do { // Leiden iteration
+        INFO(std::to_string(numberOfIterations) + " Leiden iteration(s) left");
+        numberOfIterations--;
+        changed = false;
+        bool done; // have we moved any nodes? If yes, start local moving phase again. Else stop
+        const Graph *currentGraph = G;
+        Graph coarse;
+        Partition refined;
+        calculateVolumes(*currentGraph);
+        do {
+            handler.assureRunning();
+            parallelMove(*currentGraph);
+            // If each community consists of exactly one node we're done, i.e. when |V(G)| = |P|
+            done = currentGraph->numberOfNodes() == result.numberOfSubsets();
```

Comment:
Maybe this is a bit simpler:
```suggestion
            if (currentGraph->numberOfNodes() == result.numberOfSubsets())
                break;
```